src/robin_dict.jl

```diff
@@ -0,0 +1,593 @@
+    oldsz = length(d.keys)
+    # grow at least 25%
+    if newsz < (oldsz*5)>>2
+        return d
+    end
+    rehash!(d, newsz)
+end
+
+@propagate_inbounds isslotfilled(h::RobinDict, index) = (h.hashes[index] != 0)
+@propagate_inbounds isslotempty(h::RobinDict, index) = (h.hashes[index] == 0)
+
+
+function setindex!(h::RobinDict{K,V}, v0, key0) where {K, V}
+    key = convert(K, key0)
+    isequal(key, key0) || throw(ArgumentError("$key0 is not a valid key for type $K"))
+    _setindex!(h, key, v0)
+end
+
+function _setindex!(h::RobinDict{K,V}, key::K, v0) where {K, V}
+    v = convert(V, v0)
+    sz = length(h.keys)
+    (h.count > ROBIN_DICT_LOAD_FACTOR * sz) && rehash!(h, sz<<2)
+    index = rh_insert!(h, key, v)
+    @assert index > 0
+    h.totalcost += 1
+    h
+end
+
+isempty(d::RobinDict) = (d.count == 0)
+length(d::RobinDict) = d.count
+
+"""
+    empty!(collection) -> collection
+
+Remove all elements from a `collection`.
+
+# Examples
+```jldoctest
+julia> A = RobinDict("a" => 1, "b" => 2)
+RobinDict{String,Int64} with 2 entries:
+  "b" => 2
+  "a" => 1
+
+julia> empty!(A);
+
+julia> A
+RobinDict{String,Int64} with 0 entries
+```
+"""
+function empty!(h::RobinDict{K,V}) where {K, V}
+    sz = length(h.keys)
+    empty!(h.hashes)
+    empty!(h.keys)
+    empty!(h.vals)
+    resize!(h.keys, sz)
+    resize!(h.vals, sz)
+    resize!(h.hashes, sz)
+    h.count = 0
+    h.maxprobe = 0
+    h.totalcost = 0
+    h.idxfloor = 0
+    return h
+end
+ 
+function rh_search(h::RobinDict{K, V}, key::K) where {K, V}
+    sz = length(h.keys)
+    chash = hash_key(key)
+    index = desired_index(chash, sz)
+    cdibs = 0
+    @inbounds while true
+        if isslotempty(h, index)
+            return -1
+        elseif cdibs > calculate_distance(h, index)
+            return -1
+        elseif h.hashes[index] == chash && (h.keys[index] === key || isequal(h.keys[index], key))
+            return index
+        end
+        index = (index & (sz - 1)) + 1
+    end
+end
+
+"""
+    get!(collection, key, default)
+
+Return the value stored for the given key, or if no mapping for the key is present, store
+`key => default`, and return `default`.
+
+# Examples
+```jldoctest
+julia> d = RobinDict("a"=>1, "b"=>2, "c"=>3);
+
+julia> get!(d, "a", 5)
+1
+
+julia> get!(d, "d", 4)
+4
+
+julia> d
+RobinDict{String,Int64} with 4 entries:
+  "c" => 3
+  "b" => 2
+  "a" => 1
+  "d" => 4
+```
+"""
+get!(collection, key, default)
+
+get!(h::RobinDict{K,V}, key0, default) where {K,V} = get!(()->default, h, key0)
+
+"""
+    get!(f::Function, collection, key)
+
+Return the value stored for the given key, or if no mapping for the key is present, store
+`key => f()`, and return `f()`.
+
+This is intended to be called using `do` block syntax:
+```julia
+get!(dict, key) do
+    # default value calculated here
+    time()
+end
+```
+"""
+get!(f::Function, collection, key)
+
+function get!(default::Callable, h::RobinDict{K,V}, key0::K) where {K, V}
```

Comment:
No need to ban functors

```suggestion
function get!(default, h::RobinDict{K,V}, key0::K) where {K, V}
```

Reply:
Perhaps, but that's the current API for Dict (the proposed change would introduce a method ambiguity)